xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -93,81 +95,198 @@ public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+      overallState = TRANSIENT_FAILURE;
+      startConnectionAttempt = (numConnecting == 0);
+    } else if (numConnecting > 0) {
+      overallState = CONNECTING;
+    } else if (numTransientFailure == 1 && getChildLbStates().size() > 1) {
+      overallState = CONNECTING;
+      startConnectionAttempt = true;
+    } else if (numIdle > 0) {
+      overallState = IDLE;
+    } else {
+      overallState = TRANSIENT_FAILURE;
     }
 
-    // Update the picker before shutting down the subchannels, to reduce the chance of race
-    // between picking a subchannel and shutting it down.
-    updateBalancingState();
-    for (Subchannel subchann : removedSubchannels) {
-      shutdownSubchannel(subchann);
+    RingHashPicker picker = new RingHashPicker(syncContext, ring, getImmutableChildMap());
+    // TODO(chengyuanzhang): avoid unnecessary reprocess caused by duplicated server addr updates
+    getHelper().updateBalancingState(overallState, picker);
+    this.currentConnectivityState = overallState;
+
+    // While the ring_hash policy is reporting TRANSIENT_FAILURE, it will
+    // not be getting any pick requests from the priority policy.
+    // However, because the ring_hash policy does not attempt to
+    // reconnect to subchannels unless it is getting pick requests,
+    // it will need special handling to ensure that it will eventually
+    // recover from TRANSIENT_FAILURE state once the problem is resolved.
+    // Specifically, it will make sure that it is attempting to connect to
+    // at least one subchannel at any given time.  After a given subchannel
+    // fails a connection attempt, it will move on to the next subchannel
+    // in the ring.  It will keep doing this until one of the subchannels
+    // successfully connects, at which point it will report READY and stop
+    // proactively trying to connect.  The policy will remain in
+    // TRANSIENT_FAILURE until at least one subchannel becomes connected,
+    // even if subchannels are in state CONNECTING during that time.
+    //
+    // Note that we do the same thing when the policy is in state
+    // CONNECTING, just to ensure that we don't remain in CONNECTING state
+    // indefinitely if there are no new picks coming in.
+    if (startConnectionAttempt) {
+      if (!connectionAttemptIterator.hasNext()) {
+        connectionAttemptIterator = getChildLbStates().iterator();
+      }
+      ((RingHashChildLbState)connectionAttemptIterator.next()).getLb().requestConnection();
     }
+  }
 
-    return true;
+  @Override
+  protected boolean reconnectOnIdle() {
+    return false;
+  }
+
+  @Override
+  protected boolean reactivateChildOnReuse() {
+    return false;
+  }
+
+  /**
+   * Create RingHashChildLbState objects with resolvedAddresses filled in.
+   * @return Map of {@link Endpoint} -> {@link RingHashChildLbState}
+   */
+  @Override
+  protected Map<Object, ChildLbState> createChildLbMap(ResolvedAddresses resolvedAddresses) {
```

Comment:
You probably can reduce the repeated code by just overriding the `createChildLbState()`.